### PR TITLE
feat: Add member and channel permissions calculation functions

### PR DESCRIPTION
**Pull request by mAxYoLo01**
## About

This pull request adds functions to calculate base permissions for a member, as well as calculate the member's permissions in a given channel, using permissions overwrites.

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

File: interactions/api/models/flags.py [hunk 1]
```diff
@@ -94,6 +94,11 @@ class Permissions(IntFlag):
     MODERATE_MEMBERS = 1 << 40
 
 
+ALL_PERMISSIONS = Permissions(0)
+for perm in Permissions:
+    ALL_PERMISSIONS |= perm
+
+
```
Comment (sillyElvy) [inline interactions/api/models/flags.py]: Is this an api thing?

**Review by sillyElvy — CHANGES_REQUESTED** [review]

File: interactions/api/models/flags.py (hunk 1 above)
Comment (mAxYoLo01) [reply interactions/api/models/flags.py]: This is how it is documented in the [Discord pseudocode](https://discord.com/developers/docs/topics/permissions#permission-overwrites).
Ngl I've directly copied it from Molter ext, I had it written another way which is less beautiful

**Comment by mAxYoLo01** [comment]
Since 4.3.0 has been released, I will mark this PR as ready to check.

File: interactions/api/models/channel.py [hunk 2]
```diff
@@ -1207,6 +1208,68 @@ async def join(self) -> None:
...
+        overwrite_everyone = search_iterable(
+            self.permission_overwrites, lambda overwrite: int(overwrite.id) == int(self.guild_id)
+        )
+        if overwrite_everyone:
+            permissions &= ~int(overwrite_everyone[0].deny)
+            permissions |= int(overwrite_everyone[0].allow)
+
+        # Apply role specific overwrites
+        allow = 0
+        deny = 0
```
Comment (Toricane) [inline interactions/api/models/channel.py]:
```suggestion
        allow, deny = 0, 0
```

File: interactions/api/models/channel.py [hunk 3]
```diff
@@ -1207,6 +1208,68 @@ async def join(self) -> None:
...
+                self.permission_overwrites, lambda overwrite: int(overwrite.id) == int(role_id)
+            )
+            if overwrite_role:
+                allow |= int(overwrite_role[0].allow)
+                deny |= int(overwrite_role[0].deny)
+
+        if deny:
+            permissions &= ~int(deny)
+        if allow:
+            permissions |= int(allow)
```
Comment (Toricane) [inline interactions/api/models/channel.py]: Isn't it already int?
Can you test this out pls?

```suggestion
        if deny:
            permissions &= ~deny
        if allow:
            permissions |= allow
```

**Review by Toricane — CHANGES_REQUESTED** [review]
A few small suggestions

File: interactions/api/models/channel.py (hunk 3 above)
Comment (mAxYoLo01) [reply interactions/api/models/channel.py]: You're right, relics of an older code I believe

**Review by Toricane — APPROVED** [review]
LGTM!

**Review by sillyElvy — APPROVED** [review]

**Pull request merged**